src/frontend/src/scheduler/snapshot.rs

```diff
@@ -93,25 +97,48 @@ impl std::fmt::Debug for PinnedSnapshot {
 pub type PinnedSnapshotRef = Arc<PinnedSnapshot>;
 
 impl PinnedSnapshot {
-    fn batch_query_epoch(&self, is_barrier_read: bool) -> BatchQueryEpoch {
-        let epoch = if is_barrier_read {
-            batch_query_epoch::Epoch::Current(u64::MAX)
-        } else {
-            batch_query_epoch::Epoch::Committed(self.value.max_committed_epoch)
-        };
-        BatchQueryEpoch { epoch: Some(epoch) }
+    fn batch_query_epoch(&self, scan_tables: &HashSet<TableId>) -> Result<Epoch, SchedulerError> {
+        // use the min committed epoch of tables involved in the scan
+        let epoch = scan_tables
+            .iter()
+            .map(|table_id| {
+                self.value
+                    .state_table_info
+                    .info()
+                    .get(table_id)
+                    .map(|info| Epoch(info.committed_epoch))
+                    .ok_or_else(|| anyhow!("table id {table_id} may have been dropped"))
```

Comment:
`RelationCollectorVisitor` will also add source ID to this set, and lookup join is missed in this set. I think we'd better implement a new visitor by ourselves.